Add caching for VideoPress functions with meta queries

kienstra opened the pull request:
#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Caches functions that use `WP_Query`
* Adds some arguments to `WP_Query` to improve performance, like `'no_found_rows'  => true`
* Adds a few unit tests

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Enhances an existing feature, no intended change to functionality

#### Testing instructions:
(Regression testing for VideoPress)
1. Ensure VideoPress is enabled:
![settings-jetpack](https://user-images.githubusercontent.com/4063887/75300287-04608c00-57fd-11ea-8134-362c1546c824.png)

2. Create a new post
3. Add a Video block
4. Upload a video to it
5. Expected: the video plays in the editor.
6. Expected: the video plays on the front-end

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Performance enhancement for VideoPress, caching and improving queries

kienstra committed:
Add caching for VideoPress functions that call WP_Query

These involve meta queries,
so it'd be good to have
some kind of caching.

kienstra commented on modules/videopress/class.videopress-player.php:
```diff
@@ -108,8 +108,8 @@ public function __construct( $guid, $maxwidth = 0, $options = array() ) {
 				$expire = 3600;
-				if ( isset( $video->expires ) && is_int( $video->expires ) ) {
-					$expires_diff = time() - $video->expires;
+				if ( isset( $this->video->expires ) && is_int( $this->video->expires ) ) {
```
With `unset( $video );` above, it looks like `isset( $video->expires )` will never be true

kraftbj replied:
Nice catch.

kienstra commented on modules/videopress/utility-functions.php:
```diff
@@ -655,9 +678,13 @@ function video_get_post_by_guid( $guid ) {
 
 	$query = new WP_Query( $args );
 
-	$post = $query->next_post();
```
Before, there can be an error here if the query didn't find any posts.

kienstra commented on tests/php/modules/videopress/test_functions.utility-functions.php:
```diff
@@ -12,6 +12,93 @@
+	 * this should return that instead of instantiating WP_Query.
+	 *
+	 * @covers ::video_get_post_by_guid
+	 * @since 8.3.0
```
Maybe this should be changed

jeherve replied:
Right, you'll most likely need to use `8.4.0` now, we're past the code freeze.

kienstra replied:
OK, thanks. I'll update that.

kienstra replied:
It's changed in https://github.com/Automattic/jetpack/pull/14803/commits/f91ef7cb444dcaac66470afdfd58e7c7bf738764

kienstra commented on modules/videopress/utility-functions.php:
```diff
@@ -111,7 +117,9 @@ function videopress_get_attachment_id_by_url( $url ) {
 				if ( $original_file === $file || in_array( $file, $cropped_files ) ) {
-					return (int) $attachment_id;
+					$attachment_id = (int) $attachment_id;
+					wp_cache_set( $cache_key, $attachment_id, $cache_group, HOUR_IN_SECONDS );
```
Maybe the cache expiration time should be different

jeherve commented on modules/videopress/utility-functions.php:
```diff
@@ -87,13 +85,21 @@ function videopress_get_attachment_id_by_url( $url ) {
+		$file        = basename( $url );
+		$cache_key   = 'videopress_get_attachment_id_by_url_' . md5( $url );
+		$cache_group = 'videopress';
+		$cached_id   = wp_cache_get( $cache_key, $cache_group );
```
Quick question; did you consider using transients instead of object caching? Would it be better for performance on sites that do not use caching plugins?

kienstra replied:
Good question, I'll look at that.

kienstra replied:
Sure, I could add a transient here if you'd like.

But since the attachment ID is from a  database query (WP_Query), I'm not sure that using a transient to store it in the database (wp_options) would be the best approach. 

It would probably improve performance for sites without persistent object caching. But it could fill the `wp_options` table.

dero replied:
@kienstra Also, is `videopress_get_attachment_id_by_url` ever called?

I _think_ the original usage was in `videopress_shortcode_override_for_core_shortcode` (771097dc2b1943a3095a9a34c586df85b4816a12), but has been superseded by 32dbcd7ac9ace94ce78b69831b2e35e7d55e9fa2, but the method itself has never been removed.

I ran `git log -p -S videopress_get_attachment_id_by_url` to confirm.

kienstra replied:
Uf, you're right. I missed that it's not called in Jetpack. 

The diff for that is removed in https://github.com/Automattic/jetpack/pull/14803/commits/96a565aa5734c374087016b2aec56f94d102b6e4.

kienstra replied:
Hi @jeherve, hope you had a great day.

Is there possibly another codebase where `videopress_get_attachment_id_by_url()` is called? 

If not, I could remove it here if you'd like. Though maybe that could be considered in a separate issue or PR.

jeherve replied:
It isn't used on WordPress.com right now, so it may be safe to remove. Some third party plugin / theme authors may rely on it though, so it may be best to keep it around for now, but with a `_deprecated_function` notice so we can completely remove it in a few releases.

kienstra replied:
Good idea, it's deprecated in https://github.com/Automattic/jetpack/pull/14803/commits/67df908fde1033325537cb41571048c7a4d92b4f

kraftbj replied:
I checked the VIP Go GH organization too -- no usage. That's the other big place I would expect something random to use it.

kienstra replied:
The discussion here on transients as a fallback for sites without caching is still relevant for `video_get_post_by_guid()`.

>@kienstra Expiring transients aren't autoloaded and are deleted once they expire. Do you think we need to worry about polluting wp_options under these circumstances?

It's probably not likely to fill the `wp_options` table, just on sites with a huge amount of videos and traffic.

But it seems that storing the result a database query (WP_Query) in the database isn't the best approach.

I couldn't find an example in Jetpack of storing a `WP_Query` result in a transient, though there could be examples. There are a few [non-cached $wpdb queries](https://github.com/Automattic/jetpack/blob/35c35c5a2b13446e68489210ee3edc5b12c8ed73/class.jetpack.php#L3588) stored in transients.

Still, if you have strong feelings on this, I could add transient usage. There probably isn't much risk.

leogermani replied:
Hi everyone! @jeherve asked me to share some thoughts on this discussion. 

I agree that we could use transients here. It would be more effective, since it would last through multiple requests.

Transients are not necessarily stored on the database. If memcached is enabled, for example, this option is used. And the downsize is that the expiration time is actually the maximum expiration time. So the transient may disappear earlier and we don't have much control over how much it will last. But I think this is ok for us here.

BUT, I do agree with @kienstra and I don't feel super comfortable storing the whole object in the transient. It might be harmless but if we have an edge case we are not considering here, this would make things worse.

If we would go that way, I'd say to add an extra check to see if we have a valid object before using it, just in case it got corrupted somehow.

I think a good approach here can be:

1. Use transients, but store only the post ID
2. We could use the object cache to store the post object, avoiding multiple calls to `get_post()` if we want.

And to wrap things up, I'd get the post ID in a separate method, which would be a direct replacement for the method we are deprecating :) .

kienstra replied:
>1. Use transients, but store only the post ID
>2. We could use the object cache to store the post object, avoiding multiple calls to get_post() if we want.

Thanks, what do you think about the [latest commits](https://github.com/Automattic/jetpack/compare/67df908fde1033325537cb41571048c7a4d92b4f...21dc2cdcaf16a6677e67fe155ffd3d9d68476cd5), which apply those 2 points?

kienstra replied:
>And to wrap things up, I'd get the post ID in a separate method, which would be a direct replacement for the method we are deprecating :) .

Hm, did you have in mind something like:

```php
/**
 * Using a GUID, find a post ID.
 *
 * @param string $guid The guid to use to find the post ID.
 * @return int|false The post ID.
 */
function video_get_post_id_by_guid( $guid ) {
	$post = video_get_post_by_guid( $guid );

	if ( is_object( $post ) && 'WP_Post' === get_class( $post ) ) {
		return $post->ID;
	}

	return false;
}
```

leogermani replied:
> Thanks, what do you think about the [latest commits](https://github.com/Automattic/jetpack/compare/67df908fde1033325537cb41571048c7a4d92b4f...21dc2cdcaf16a6677e67fe155ffd3d9d68476cd5), which apply those 2 points?

What I thought is that we can use both transients and wp_cache at the same time, each one for a purpose. Transients to store the post ID, so we avoid making the costly SQL SELECT many times though several requests. Object Caching to avoid calling `get_post()` several times within the same request.

> Hm, did you have in mind something like:

Not exactly. I would actually leave the `WP_Query` call in `video_get_post_id_by_guid`, adding the `'fields' => 'ids'`. I would call it from `video_get_post_by_guid` and then call `get_post`. So I would have the same behavior every time. (Always fetch the ID and then call `get_post`)

Looks a bit cleaner to me. Each method doing just one thing, and avoiding things like `$cached_post` sometimes holding the actual post, and sometimes holding just the post id.

Also it's important to add an expiration time for the transient.

This is how I see it. (Please consider this just as an opinion!)

```PHP

function video_get_post_by_guid( $guid ) {
	// ...
	$cached_post = wp_cache_get( $cache_key, $cache_group );

	if ( $cached_post ) {
		return $cached_post;
	}

	$post_id = $this->video_get_post_id_by_guid( $guid );

	if ( is_int( $post_id ) ) {
		$post = get_post( $cached_post );
		wp_cache_set( $cache_key, $post, $cache_group, HOUR_IN_SECONDS );
		if ( $post ) {
			return $post;
		}
	}

}

function video_get_post_id_by_guid( $guid ) {
	// ... 
	$cached_id = get_transient( $cache_key );
	if ( $cached_id ) {
		return $cached_id;
	}

	// do the query
	set_transient( $cache_key, $id, HOUR_IN_SECONDS );
	return $id;
}
```

kienstra replied:
Thanks, and sorry for the delay. I'm working on this.

kienstra replied:
How does [this look](https://github.com/Automattic/jetpack/compare/21dc2cdcaf16a6677e67fe155ffd3d9d68476cd5...331f265212a39d91d1447f4d316ed6374fc5b3b6)? It mainly uses your [snippet above](https://github.com/Automattic/jetpack/pull/14803#discussion_r393672549).

kienstra committed:
Change the since tag to 8.4.0

As Jeremy mentioned,
8.3.0 now has a code freeze.

kienstra commented:
Hi @dero,
Hope you're doing great.

Could you please review this?

dero reviewed (commented):
@kienstra Changes look good. I had a question about transients vs. wp_cache_* usage and I also think one of the methods that receives caching is no longer used in the codebase and can be removed.

kienstra committed:
Revert changes to videopress_get_attachment_id_by_url()

As Dero mentioned,
it looks like this isn't used in Jetpack.

kienstra committed:
Deprecate videopress_get_attachment_id_by_url()

As Jeremy mentioned, this should
give third-party developers a chance
to change to something else.

kraftbj commented:
Marking for author reply pending resolution of the `wp_cache_*` vs transient question.

kienstra commented:
Thanks, I [responded above](https://github.com/Automattic/jetpack/pull/14803#discussion_r389323697).

leogermani commented on modules/videopress/utility-functions.php:
```diff
@@ -73,13 +73,16 @@ function videopress_get_video_details( $guid ) {
  *
  * Modified from https://wpscholar.com/blog/get-attachment-id-from-wp-image-url/
  *
+ * @deprecated since 8.4.0
```
Not required, but nice if we added a `@see` keyword pointing to a close replacement (https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#1-2-deprecated-functions)

kienstra replied:
Good idea.

Though I couldn't find a possible replacement.

This looks to do the opposite:
https://github.com/Automattic/jetpack/blob/91acb92673f2c44b12cbdac18fb26aa966776369/modules/videopress/utility-functions.php#L663-L671

leogermani replied:
Now we have one!

`video_get_post_id_by_guid`

kienstra replied:
OK, thanks. Applied in https://github.com/Automattic/jetpack/pull/14803/commits/5c05e33d54e8a7987024dc3c7891f54ce6c5a572

leogermani commented on modules/videopress/utility-functions.php:
```diff
@@ -73,13 +73,16 @@ function videopress_get_video_details( $guid ) {
  * Modified from https://wpscholar.com/blog/get-attachment-id-from-wp-image-url/
  *
+ * @deprecated since 8.4.0
  * @todo: Add some caching in here.
```
We could remove it :)

kienstra replied:
Ah, OK. Remove the entire function `videopress_get_attachment_id_by_url()` ?

leogermani replied:
No, I mean remove this "@todo" comment, sorry

kienstra replied:
Ah, good idea.

kienstra replied:
Removed in https://github.com/Automattic/jetpack/commit/85b14a0d38bbdddbcb7a1ba88e968587546848b5#diff-66c3e0655a6e9d684e7898c2c70171c3L77

kienstra committed:
Store the videopress ID in a transient if no persistent object caching

Also, store the full WP_Post in the cache,
if it is available.

kienstra committed:
Remove needless helper method

There is actually an existing function
that does the same thing.

kienstra committed:
Add a new function, using Leo's suggestion

Add video_get_post_id_by_guid(),
and a unit test for it.

kienstra committed:
Move the new function's test lower in the file

To correspond to its position
in utility-functions.php.

kienstra committed:
Set the transient expiration to an hour

Following Leo's snippet,
this should have an expiration.

leogermani commented on tests/php/modules/videopress/test_functions.utility-functions.php:
```diff
@@ -12,6 +12,181 @@
+	}
+
+	/**
+	 * Gets the test data for test_cached_video_get_post_by_guid().
```
minor: I think you meant `test_non_cached_video_get_post_by_guid`

kienstra replied:
Sure, good catch. Applied in https://github.com/Automattic/jetpack/pull/14803/commits/5c05e33d54e8a7987024dc3c7891f54ce6c5a572

leogermani commented on modules/videopress/utility-functions.php:
```diff
@@ -637,13 +639,48 @@ function video_image_url_by_guid( $guid, $format ) {
 		'post_mime_type' => 'video/videopress',
 		'post_status'    => 'inherit',
+		'no_found_rows'  => true,
 		'meta_query'     => array(
```
As we are only using the ID here, we can use the ` 'fields' => 'ids' ` param in the query to SELECT only the ID from the database.

kienstra replied:
Ah, sorry for missing that. It's added in https://github.com/Automattic/jetpack/commit/64918b1371d1dde532e7bd2e5f467d84ba3f32e4#diff-66c3e0655a6e9d684e7898c2c70171c3R685

leogermani reviewed (changes requested):
Hi @kienstra ,

Thanks a lot for the changes.

There are few small things though:

- [x] Use the `fields => ids` param in the query. Since this PR is all about performance, this is a must-have.
- [x] Add the `@since` keywork to the new functions. (see https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#1-functions-class-methods)
- [x] Now we have a perfect replacement for that deprecated function! Add there in the `@see` keyword there with our new function.
- [x] I see that this file have functions prefixed both with `video` and with `videopress`. Most of them with `videopress` and, since this is the name of the module, we should stick to it. Those functions that are using only `video` as prefix we can review another time. But I think we should enforce the most appropriate prefix. So, change the prefix of the new functions to `videopress`

kienstra committed:
Add a see tag, correct DocBlock summary

The see tag points to a possible replacement.
Also, the DocBlock summary references the wrong function.

kienstra committed:
Change prefix of new function from video_ to videopress_

Also change the transient cache key.
Also, simply return the post, instead of
accessing the ->ID property.

kienstra commented:
>I see that this file have functions prefixed both with video and with videopress. Most of them with videopress and, since this is the name of the module, we should stick to it. Those functions that are using only video as prefix we can review another time. But I think we should enforce the most appropriate prefix. So, change the prefix of the new functions to videopress

OK, is https://github.com/Automattic/jetpack/pull/14803/commits/64918b1371d1dde532e7bd2e5f467d84ba3f32e4 what you had in mind? It renames `video_get_post_id_by_guid` to `videopress_get_post_id_by_guid()`

kienstra committed:
Add since tags to new functions

videopress_get_post_id_by_guid()
now has a @since tag,
in addition to other dataProviders

kienstra committed:
Remove needless comment

This applied when it was part of
video_get_post_by_guid(),
but not to the new function it's in.

kienstra commented:
Hi @leogermani,
Thanks for your review. All 4 of [your points](https://github.com/Automattic/jetpack/pull/14803#pullrequestreview-377774802) are applied.

leogermani committed:
Rename function to keep consistency

leogermani reviewed (approved):
Hi @kienstra ,

Thanks for all the work, it looks great now!

I had a quick chat with the team and we thought it was worth to rename one of the old functions to have a bit more consistency. I already did it since it doesn't change anything with the logic you implemented.

kienstra commented:
>I had a quick chat with the team and we thought it was worth to rename one of the old functions to have a bit more consistency. I already did it since it doesn't change anything with the logic you implemented.

Sounds good, thanks for pushing that change.
